File: firmware/src/motor.cpp
#include "motor.h"
#include "pio_rotary_encoder.h"

Motor::Motor(
  bool isLeft,
  Store* store,
  RotaryEncoder* encoder,
  int in1,
  int in2,
  int pwm,
  int stby
) {
  _isLeft = isLeft;
  _store = store;
  _in1 = in1;
  _in2 = in2;
  _stby = stby;
  _encoder = encoder;
  _pwm = pwm;
}

void Motor::begin() {
  pinMode(_in1, OUTPUT);
  pinMode(_in2, OUTPUT);
  analogWriteFreq(1000);
  analogWriteRange(65535);

  pinMode(_stby, OUTPUT);
  digitalWrite(_stby, LOW);
  pinMode(_pwm, OUTPUT);

  _pos = _store->getLastPosition();
  setpoint = _pos;
  _encoder->set_rotation(0);
  resetPid();
}

int32_t Motor::getPos() {
  return _pos;
}

void Motor::resetPid() {
  speed_previous_error = 0;
  PID_i = 0;
  last_update_time_ms = millis();
  last_non_goal_update_ms = millis();
  last_position_update = getPos();
}

int32_t Motor::getSlowdownSteps() {
  int32_t limit = _store->getBottomLimit();
  float percent = float(_store->getSlowdownPercent()) / 100;
  return limit * percent;
}

void Motor::goToPositionPercent(uint8_t positionPercent) {
  int32_t limit = _store->getBottomLimit();
  float percent = positionPercent / 100.0;
  setpoint = limit * percent;
  Serial.printf("Going to position percent %d, pos: %d, setpoint %d, limit %d\n",
    positionPercent,
    getPos(),
    setpoint,
    limit
  );
  int stepsLeft = abs(setpoint - getPos());

  int slowdownSteps = getSlowdownSteps();
  if (stepsLeft <= slowdownSteps) {
    int approximatePwmSpeed = calculateSlowdownSpeed(stepsLeft, slowdownSteps, 100);
    pwm_speed = getPos() < setpoint ? approximatePwmSpeed : -approximatePwmSpeed;
    max_speed_abs = _store->getDefaultMaxSpeed();
  } else {
    max_speed_abs = -1;
  }

  goal_reached_at = -1;
  goal_reached = false;
}

float Motor::calculateSlowdownSpeed(int32_t stepsLeft, int32_t slowdownSteps, float maxSpeed) {
  if (stepsLeft > slowdownSteps) {
    return maxSpeed;
  }

  float leftRatio = float(stepsLeft) / slowdownSteps;
  return maxSpeed * leftRatio;
}

void Motor::ccw(float duty) {
  digitalWrite(_stby, HIGH);
  digitalWrite(_in1, HIGH);
  digitalWrite(_in2, LOW);
  int duty_16 = int((duty * 65536) / 100);
  analogWrite(_pwm, duty_16);
}
void Motor::cw(float duty) {
  digitalWrite(_stby, HIGH);
  digitalWrite(_in1, LOW);
  digitalWrite(_in2, HIGH);
  int duty_16 = int((duty * 65536) / 100);
  analogWrite(_pwm, duty_16);
}
void Motor::short_break() {
  digitalWrite(_stby, HIGH);
  digitalWrite(_in1, HIGH);
  digitalWrite(_in2, HIGH);
  analogWrite(_pwm, 65536);
}
void Motor::stop() {
  digitalWrite(_in1, LOW);
  digitalWrite(_in2, LOW);
  analogWrite(_pwm, 0);
}

void Motor::up(float speed) {
  if (!_isLeft) {
    ccw(speed);
  } else {
    cw(speed);
  }
}
void Motor::down(float speed) {
  if (!_isLeft) {
    cw(speed);
  } else {
    ccw(speed);
  }
}

void Motor::setSpeed(float speed) {
  pwm_speed = speed;

  if (speed == 0) {
    short_break();
  } else if (speed < 0) {
    up(abs(speed));
  } else if (speed > 0) {
    down(abs(speed));
  }
}

void Motor::goSetupUp() {
  is_setup = true;
  up(_store->getManualSpeedUp());
}

void Motor::stopAndSetTopLimit() {
  setSpeed(0);
  _pos = 0;
  _last_updated_pos = 0;
  _encoder->set_rotation(0);
  _store->storeLastPosition(0);
}

void Motor::goSetupDown() {
  is_setup = true;
  down(_store->getManualSpeedDown());
}

void Motor::stopAndFinishSetup() {
  _store->storeBottomLimit(getPos());
  is_setup = false;
  is_measuring_default_speed = true;
  goToPositionPercent(50);
}

void Motor::storeLastPositionUpdate() {
  if (last_position_store_update_time_ms == 0) {
    last_position_store_update_time_ms = millis();
    return;
  }

  if (millis() - last_position_store_update_time_ms > 1000) {
    last_position_store_update_time_ms = millis();
    _store->storeLastPosition(getPos());
  }
}

bool Motor::isGoalReached() {
  if (goal_reached) return true;

  float tolerance = _store->getBottomLimit() * 0.005;
  // int bottomLimit = _store->getBottomLimit();
  // Serial.print("Tolerance: ");
  // Serial.print(tolerance, 2);
  // Serial.printf(", bottom limit: %d, diff: %d\n", bottomLimit, abs(_pos - setpoint));

  // TODO reenable tolerance
  if (abs(_pos - setpoint) <= tolerance) {
  //if (abs(getPos() - setpoint) == 0) {
    if (goal_reached_at != -1 && millis() - goal_reached_at > 1000) {
      stop();
      goal_reached = true;
      Serial.printf("GOAL stabilised");
    } else {
      short_break();
      if (goal_reached_at == -1) {
        goal_reached_at = millis();
      }
    }

    return true;
  }

  return false;
}

void Motor::updatePidSpeed(long timeDiffMs, float actualSpeed, float desiredSpeed) {
  float speedError = desiredSpeed - actualSpeed;
  float PID_p = _store->getPidKp() * speedError;
  float PID_d = speed_previous_error != 0 ? _store->getPidKd() * (float(speedError - speed_previous_error) / timeDiffMs) : 0;
  PID_i = PID_i + (_store->getPidKi() * speedError);

  speed_previous_error = speedError;
  float PID_total = PID_p + PID_i + PID_d;

  float calculatedPwmSpeed;
  if (PID_total > 0) {
    float totalAdjustment = min(5, PID_total);
    calculatedPwmSpeed = min(100, pwm_speed + totalAdjustment);
  } else {
    float totalAdjustment = max(-5, PID_total);
    calculatedPwmSpeed = max(-100, pwm_speed + totalAdjustment);
  }

  // Serial.printf("%d,%d,", getPos(), setpoint);
  // Serial.print(desiredSpeed, 4);
  // Serial.print(",");
  // Serial.print(actualSpeed, 4);
  // Serial.print(",");
  // Serial.print(pwm_speed);
  // Serial.print(",");
  // Serial.print(calculatedPwmSpeed, 2);
  // Serial.print(",");
  // Serial.print(PID_total, 2);
  // Serial.print(",");
  // Serial.print(PID_p, 2);
  // Serial.print(",");
  // Serial.print(PID_i, 2);
  // Serial.print(",");
  // Serial.print(PID_d, 2);
  // Serial.println("");

  setSpeed(calculatedPwmSpeed);
}

void Motor::updatePos() {
  unsigned long time = millis();
  if (time - last_pos_update_time_ms > 5) {
    last_pos_update_time_ms = time;
    int current_pos = _isLeft ? -_encoder->get_rotation() : _encoder->get_rotation();
    int diff = current_pos - _last_updated_pos;
    _pos = _pos + diff;
    
    _last_updated_pos = current_pos;
  }
}

void Motor::update() {
  updatePos();
  storeLastPositionUpdate();
  if (is_setup) return;
  if (isGoalReached()) return;

  if (last_update_time_ms == 0) {
    last_update_time_ms = millis();
  }

  unsigned long time_diff_ms = millis() - last_update_time_ms;
  if (time_diff_ms < 40) return;

  last_update_time_ms = millis();

  int position_diff = getPos() - last_position_update;

  last_position_update = getPos();

  // once moved 10% up
  if (is_measuring_default_speed && float(getPos()) / _store->getBottomLimit() < 0.9) {
    is_measuring_default_speed = false;
    float actualSpeed = abs(float(position_diff) / time_diff_ms);
    Serial.println("Storing default max speed");
    _store->storeDefaultMaxSpeed(actualSpeed);
  }

  int steps_left = setpoint - getPos();

  if (abs(steps_left) <= getSlowdownSteps()) {
    float actualSpeed = float(position_diff) / time_diff_ms;
    
    // after switching from 100% speed to slowdown set max speed to the current one for smooth transition
    if (max_speed_abs == -1) {
      max_speed_abs = abs(actualSpeed);
    }
    float desiredSpeed = calculateSlowdownSpeed(abs(steps_left), getSlowdownSteps(), max_speed_abs);

    // Serial.print("Actual speed ");
    // Serial.print(actualSpeed, 2);
    //Serial.print(", Desired Speed ");
    //Serial.print(desiredSpeed, 2);
    // Serial.print(", Max speed ");
    // Serial.print(max_speed_abs, 2);      
    //Serial.println("");

    // getPos() < setpoint ? _store->getDefaultMaxSpeed() : -_store->getDefaultMaxSpeed();
    updatePidSpeed(time_diff_ms, actualSpeed, getPos() < setpoint ? desiredSpeed : -desiredSpeed);
  } else {
    setSpeed(getPos() < setpoint ? 100 : -100);
  }
}

int Motor::writeStatus(uint8_t* buffer) {
  int32_t pos = getPos();
  memcpy(&buffer[0], &pos, 4);
  memcpy(&buffer[4], &setpoint, 4);
  buffer[8] = goal_reached;
  return 9;
}

File: firmware-micropython/app/motor.py
from machine import Pin, PWM
from utime import sleep
import time
import _thread
import micropython
micropython.alloc_emergency_exception_buf(100)

class Motor:
    
    def __init__(self, is_left, store, pins):
        self._is_left = is_left
        self._store = store
        
        self._in1 = Pin(pins['in1'],Pin.OUT)
        self._in2 = Pin(pins['in2'], Pin.OUT)
        self._pwm = PWM(Pin(pins['pwm']))
        self._pwm.freq(1000)
        self._stby = Pin(pins['stby'], Pin.OUT)
        self._stby.value(1)
        
        self._pos = self._store.get_last_position()

        self._enc2 = Pin(pins['enc2'], Pin.IN)
        self._enc1 = Pin(pins['enc1'], Pin.IN)
        self._enc1.irq(trigger = Pin.IRQ_RISING, handler=self.handle_enc1, hard = True)

        self._going_down = False

        self._state = 0
        self._last_update_time_ms = None
        self._last_position_store_update_time_ms = None
        self._last_non_goal_update = None
        
        self._MAX_PWM = 100
        
        self._is_setup = False
        self._is_measuring_default_speed = False
        
        # when powered on we assume that it's the desired goal position
        self._goal_reached_at = None
        self._goal_reached = False
        self._setpoint = self._pos

        self._last_position_update = None
        self._speed_previous_error = 0.0
        self._PID_i = 0.0
        self._pwm_speed = 0
        self._max_speed = None

        self._reset_pid()

        self._direction_changes = 0
        self._false_triggers = 0

    def _reset_pid(self):
        self._speed_previous_error = 0.0
        self._PID_i = 0.0
        self._last_update_time_ms = time.ticks_ms()
        self._last_non_goal_update = time.ticks_ms()
        self._last_position_update = self._pos

    def _get_slowdown_steps(self):
        limit = self._store.get_bottom_limit()
        return round((limit * self._store.get_slowdown_percent())/100)

    def handle_enc1(self, pin):
        if pin.value() != 1:
            self._false_triggers += 1

        if self._enc2.value() == 1:
            if not self._going_down:
                self._direction_changes += 1
            self._going_down = True    
            self._pos -= 1 if self._is_left else -1
        else:
            if self._going_down:
                self._direction_changes += 1
            self._going_down = False
            self._pos += 1 if self._is_left else -1

    def go_to_position_percent(self, position_percent):
        limit = self._store.get_bottom_limit()
        position = round((limit * position_percent)/100)
        #+print(f'going to position {position}, slowdown steps: {self._get_slowdown_steps()}')
        self._setpoint = position

        steps_left = abs(position - self._pos)
        if steps_left <= self._get_slowdown_steps():
            #+print("Position difference is shorter than slowdown period, using default values for pwm speed and max speed")
            approximate_pwm_speed = self._calculate_speed(steps_left, self._get_slowdown_steps(), self._MAX_PWM)
            self._pwm_speed = approximate_pwm_speed if self._pos < self._setpoint else -approximate_pwm_speed
            self._max_speed = self._store.get_default_max_speed() if self._pos < self._setpoint else -self._store.get_default_max_speed()              
        else:
            self._max_speed = None
        
        self._goal_reached_at = None
        self._goal_reached = False
        
    def go_setup_up(self):
        self._is_setup = True
        speed = self._MAX_PWM * self._store.get_manual_speed_up() / 100
        #+print(f'Go setup up with pwm: {speed}')
        self._direction_changes = 0
        self.up(speed)

    def stop_and_set_top_limit(self):     
        self._set_speed(0)
        self._pos = 0
        self._store.store_last_position(0)
        print(f'stop and set top limit, direction changes: {self._direction_changes}')
        self._direction_changes = 0

    def go_setup_down(self):
        self._is_setup = True
        speed = self._MAX_PWM * self._store.get_manual_speed_down() / 100
        #+print(f'Go setup down with pwm: {speed}')
        self.down(speed)

    def finish_bottom_setup(self):
        self._store.store_bottom_limit(self._pos)
        self._is_setup = False
        self._is_measuring_default_speed = True
        self._direction_changes = 0
        self.go_to_position_percent(50)    

    def _calculate_speed(self, steps_left, slowdown_steps, max_speed): 
        if steps_left > slowdown_steps:
            return max_speed
        
        left_ratio = steps_left/slowdown_steps
        return max_speed * left_ratio

    def _store_last_position_update(self):
        if self._last_position_store_update_time_ms is None:
            self._last_position_store_update_time_ms = time.ticks_ms()
            return
        time_diff_ms = time.ticks_diff(time.ticks_ms(), self._last_position_store_update_time_ms)  
        if time_diff_ms > 1000:
            #+print(f'Pos: {self._pos}, direction changes: {self._direction_changes}')
            self._last_position_store_update_time_ms = time.ticks_ms()

            # don't wirte to FS while motor is moving to avoid interrupt delays
            if self._goal_reached and not self._is_setup:
                self._store.store_last_position(self._pos)

    def format_status(self):
        return f"{self._pos}:{self._setpoint}:{1 if self._goal_reached else 0}"

    def _check_goal_reach(self):
        if self._goal_reached == True:
            return True

        # 1% tolerance for reaching the goal
        tolerance = self._store.get_bottom_limit() * 0.005

        if abs(self._pos - self._setpoint) <= tolerance:
            if self._goal_reached_at is not None and time.ticks_diff(time.ticks_ms(), self._goal_reached_at) > 1000:
                #+print('GOAL STABILISED')
                self.stop()
                self._goal_reached = True
                print(f'GOAL STABILISED, direction changes: {self._direction_changes}')
            else:
                self.short_break()
                if self._goal_reached_at is None:
                  self._goal_reached_at = time.ticks_ms()
            
            return True

        return False

    def update(self):        
        self._store_last_position_update()

        if self._is_setup:
            return

        if self._check_goal_reach():
            return

        if self._last_update_time_ms is None:
            self._last_update_time_ms = time.ticks_ms()                
        time_diff_ms = time.ticks_diff(time.ticks_ms(), self._last_update_time_ms)  
        
        if time_diff_ms < 40:
            return

        position_diff = self._pos - self._last_position_update
        
        self._last_update_time_ms = time.ticks_ms()
        self._last_position_update = self._pos

        # once we moved 10% up
        if self._is_measuring_default_speed == True and self._pos/self._store.get_bottom_limit() < 0.9:
            self._is_measuring_default_speed = False
            actual_speed = abs(position_diff / time_diff_ms)
            self._store.store_default_max_speed(actual_speed)

        # going down
        if self._pos < self._setpoint:
            steps_left = self._setpoint - self._pos
            if steps_left <= self._get_slowdown_steps():
                actual_speed = position_diff / time_diff_ms
                if self._max_speed is None:
                    self._max_speed = actual_speed
                desired_speed = self._calculate_speed(steps_left, self._get_slowdown_steps(), self._max_speed)
                #+print(f'steps left: {steps_left}, desired speed: {desired_speed}, actual_speed: {actual_speed}, max speed: {self._max_speed}')
                self._update_pid_speed(time_diff_ms, actual_speed, desired_speed)
            else:
                self._set_speed(self._MAX_PWM)
                
        # going up
        elif self._pos > self._setpoint:
            steps_left = self._pos - self._setpoint
            if steps_left <= self._get_slowdown_steps():
                actual_speed = position_diff / time_diff_ms
                if self._max_speed is None:
                    self._max_speed = actual_speed     
                desired_speed = self._calculate_speed(steps_left, self._get_slowdown_steps(), self._max_speed)
                #+print(f'steps left: {steps_left}, desired speed: {desired_speed}, actual_speed: {actual_speed}, max speed: {self._max_speed}')
                self._update_pid_speed(time_diff_ms, actual_speed, desired_speed)
            else:
                self._set_speed(-self._MAX_PWM)

    def _update_pid_speed(self, time_diff_ms, actual_speed, desired_speed):        
        speed_error = desired_speed - actual_speed
        PID_p = self._store.get_pid_kp() * speed_error
        
        #dist_diference = speed_error - self._speed_previous_error
        PID_d = self._store.get_pid_kd() * ((speed_error - self._speed_previous_error)/time_diff_ms) if self._speed_previous_error != 0.0 else 0
        self._PID_i = self._PID_i + (self._store.get_pid_ki() * speed_error)
        
        #print(f'position travelled {position_diff}, actual_speed: {actual_speed}, pid_p: {PID_p}, pid_i: {self._PID_i} pwm_speed: {self._pwm_speed}')
        
        self._speed_previous_error = speed_error

        PID_total = (PID_p + self._PID_i + PID_d) * (self._MAX_PWM/100)
        #+print(f'PID_total: {PID_total}, {PID_p} + {self._PID_i} + {PID_d}')
        
        if PID_total > 0:
            total_adjustment = min(5, PID_total)
            pwm_speed = min(self._MAX_PWM, self._pwm_speed + total_adjustment)
        else:
            total_adjustment = max(-5, PID_total)
            pwm_speed = max(-self._MAX_PWM, self._pwm_speed + total_adjustment)
            
        new_speed = self._pwm_speed + total_adjustment    

        if new_speed > 0:
            pwm_speed = min(self._MAX_PWM, new_speed)
        else:
            pwm_speed = max(-self._MAX_PWM, new_speed)

        self._set_speed(pwm_speed)        


    def _set_speed(self, speed):
        #+print(f'Setting speed to {speed}')
        self._pwm_speed = speed
        if speed < 0:
            self.up(abs(speed))
        elif speed > 0:
            self.down(abs(speed))
        elif speed == 0:
            self.short_break()

    def up(self, speed):
        if not self._is_left:
            self._ccw(speed)
        else:
            self._cw(speed)
            
    def down(self, speed):
        if not self._is_left:
            self._cw(speed)
        else:
            self._ccw(speed)        

    def _ccw(self, duty):
        self._stby.value(1)
        self._in1.value(1)
        self._in2.value(0)
        duty_16 = int((duty*65536)/100)
        self._pwm.duty_u16(duty_16)

    def _cw(self, duty):
        self._stby.value(1)
        self._in1.value(0)
        self._in2.value(1)
        duty_16 = int((duty*65536)/100)
        self._pwm.duty_u16(duty_16)
        
    def stop(self):
        self._in1.value(0)
        self._in2.value(0)
        self._pwm.duty_u16(0)
        #self._stby.value(0)

    def short_break(self):
        self._stby.value(1)
        self._in1.value(1)
        self._in2.value(1)
        duty_16 = int(65536)
        self._pwm.duty_u16(duty_16)